pkg/build/controller/common/util.go

```diff
@@ -134,3 +138,43 @@ func HasBuildPodNameAnnotation(build *buildapi.Build) bool {
 	_, hasAnnotation := build.Annotations[buildapi.BuildPodNameAnnotation]
 	return hasAnnotation
 }
+
+// ProcessValueFrom supports the k8s pod syntax of allowing "valueFrom" in addition
+// to "value", where "valueFrom" uses the downward api to reference existing env
+// variables or jsonpath references to the build object.
+func ProcessValueFrom(build *buildapi.Build, client kclientset.Interface) (*[]kapi.EnvVar, *[]kapi.EnvVar, error) {
+	var outputEnv []kapi.EnvVar
+	var allErrs []error
+
+	inputEnv := buildutil.GetEnvFromStrategy(build)
+	store := envutil.NewResourceStore()
+
+	for _, e := range *inputEnv {
+		var value string
+		var err error
+
+		switch {
+		case e.Value != "":
+			value = e.Value
+		case e.ValueFrom.FieldRef != nil:
+			value, err = envutil.GetEnvVarFieldRef(build, e.ValueFrom)
+		case e.ValueFrom.SecretKeyRef != nil:
+			value, err = envutil.GetEnvVarSecretRefValue(client, build.Namespace, store, e.ValueFrom.SecretKeyRef)
+		case e.ValueFrom.ConfigMapKeyRef != nil:
+			value, err = envutil.GetEnvVarConfigMapRefValue(client, build.Namespace, store, e.ValueFrom.ConfigMapKeyRef)
+		}
+
+		if err != nil {
+			allErrs = append(allErrs, err)
+			continue
+		}
+
+		outputEnv = append(outputEnv, kapi.EnvVar{Name: e.Name, Value: value})
+	}
+
+	if len(allErrs) != 0 {
+		return nil, nil, utilerrors.NewAggregate(allErrs)
+	}
+
+	return inputEnv, &outputEnv, nil
```

Comment:
i think i see why you did it, but i think it's weird you're returning the inputenv here.  i think you should just return the resolved env.